File: python/analysis/scoring.py
def calculate_score(results: dict) -> int:
    """Calculate overall content quality score (0–100)."""
    try:
        score = 100

        # Content depth
        depth = results.get('content_depth', {}).get('depth_label', 'thin')
        if depth == 'thin':
            score -= 15
        elif depth == 'short':
            score -= 7

        # Content relevance
        relevance = results.get('content_relevance', {})
        if not relevance.get('title_in_h1', False):
            score -= 10
        if not relevance.get('title_in_intro', False):
            score -= 5
        if relevance.get('heading_alignment_score', 1.0) < 0.5:
            score -= 5
        if relevance.get('keyword_stuffing_detected', False):
            score -= 15

        # E-E-A-T
        eeat_count = results.get('eeat_signals', {}).get('eeat_signals_count', 0)
        if eeat_count == 0:
            score -= 20
        elif eeat_count == 1:
            score -= 12
        elif eeat_count == 2:
            score -= 6

        # Freshness
        freshness = results.get('content_freshness', {}).get('freshness_status', 'undated')
        if freshness == 'very_stale':
            score -= 10
        elif freshness == 'stale':
            score -= 5

        # Thin content
        thin = results.get('thin_content', {}).get('thin_content_risk', 'none')
        if thin == 'high':
            score -= 20
        elif thin == 'medium':
            score -= 10
        elif thin == 'low':
            score -= 5

        # Anchor text
        anchor = results.get('anchor_text_quality', {}).get('anchor_quality_score', 'excellent')
        if anchor == 'poor':
            score -= 10
        elif anchor == 'fair':
            score -= 5

        # Readability
        reading_level = results.get('readability', {}).get('reading_level', 'moderate')
        if reading_level == 'academic':
            score -= 10
        elif reading_level == 'difficult':
            score -= 5

        return max(0, min(100, score))
    except Exception:
        return 0


def get_score_label(score: int) -> str:
    """Return human-readable label for score."""
    if score >= 80:
        return 'excellent'
    elif score >= 60:
        return 'good'
    elif score >= 40:
        return 'needs_work'
    else:
        return 'poor'


def collect_all_issues(results: dict) -> list:
    """Collect human-readable issue strings from all categories."""
    issues = []
    try:
        # Depth
        depth_issues = results.get('content_depth', {}).get('issues', [])
        issues.extend(depth_issues)

        depth_label = results.get('content_depth', {}).get('depth_label', '')
        if depth_label == 'thin':
            issues.append('Thin content: page has fewer than 100 words')
        elif depth_label == 'short':
            issues.append('Short content: page has fewer than 300 words')

        # Relevance
        relevance = results.get('content_relevance', {})
        if not relevance.get('title_in_h1', True):
            issues.append('Title keywords not found in H1 heading')
        if not relevance.get('title_in_intro', True):
            issues.append('Title keywords not found in page introduction')
        if relevance.get('heading_alignment_score', 1.0) < 0.5:
            issues.append('Low heading alignment: headings do not reflect content below them')
        if relevance.get('keyword_stuffing_detected', False):
            issues.append('Keyword stuffing detected: a single term exceeds 5% of all words')

        # E-E-A-T
        eeat_label = results.get('eeat_signals', {}).get('eeat_label', 'strong')
        if eeat_label == 'weak':
            issues.append('Weak E-E-A-T signals: add first-person experience, statistics, or citations')

        eeat = results.get('eeat_signals', {})
        if eeat.get('time_sensitive_without_date', False):
            issues.append('Time-sensitive phrases found without a nearby date reference')

        # Freshness
        freshness = results.get('content_freshness', {}).get('freshness_status', '')
        if freshness == 'very_stale':
            issues.append('Content appears very outdated: most recent year is 3+ years ago')
        elif freshness == 'stale':
            issues.append('Content may be stale: most recent year is 2 years ago')

        # Thin content
        thin = results.get('thin_content', {})
        if thin.get('boilerplate_present', False):
            issues.append(f"Boilerplate text detected: {', '.join(thin.get('boilerplate_detected', []))}")
        if thin.get('high_repetition', False):
            issues.append('High sentence repetition detected: possible auto-generated content')
        if thin.get('skeleton_page_detected', False):
            issues.append('Skeleton page: more headings than paragraphs — add substantive content')

        # Anchor quality
        anchor_score = results.get('anchor_text_quality', {}).get('anchor_quality_score', 'excellent')
        if anchor_score == 'poor':
            issues.append('Poor anchor text quality: most links use generic phrases like "click here"')
        elif anchor_score == 'fair':
            issues.append('Fair anchor text quality: consider using more descriptive link text')

        # Readability
        readability = results.get('readability', {})
        reading_level = readability.get('reading_level', 'moderate')
        if reading_level == 'academic':
            issues.append(f'Academic reading level (score: {readability.get("flesch_reading_ease", "n/a")}): content may be too complex for a general audience')
        elif reading_level == 'difficult':
            issues.append(f'Difficult reading level (score: {readability.get("flesch_reading_ease", "n/a")}): consider simplifying sentences')
        long_s = readability.get('long_sentences_count', 0)
        if long_s > 3:
            issues.append(f'{long_s} sentences over 30 words — consider breaking them up')

        # TOC recommendation
        toc = results.get('toc', {})
        if toc.get('toc_recommended', False):
            issues.append('Long-form content (>1500 words) has no table of contents — consider adding one')

        # Heading hierarchy
        hierarchy = results.get('heading_hierarchy', {})
        if not hierarchy.get('hierarchy_valid', True):
            violations = hierarchy.get('violations', [])
            issues.append(f'Heading hierarchy violations: {len(violations)} level skip(s) detected')

        # Meta description
        meta_desc = results.get('meta_description', {})
        meta_status = meta_desc.get('meta_description_status', '')
        if meta_status == 'missing':
            issues.append('Meta description is missing')
        elif meta_status == 'too_short':
            issues.append(f'Meta description is too short ({meta_desc.get("meta_description_length")} chars, min 120)')
        elif meta_status == 'too_long':
            issues.append(f'Meta description is too long ({meta_desc.get("meta_description_length")} chars, max 160)')

        # Image alt text
        img = results.get('image_alt_text', {})
        if img.get('images_missing_alt', 0) > 0:
            issues.append(f'{img["images_missing_alt"]} image(s) missing alt text')

        # Transition words
        tw = results.get('transition_words', {})
        if tw.get('transition_label') == 'low':
            issues.append('Low transition word usage: consider adding more connective phrases')

    except Exception:
        pass

    return issues

File: src/analysis/scoring.ts
import { FetchResult } from './fetch';
import { FieldData, LabData } from './psi';

/**
 * DOM structure analysis from Python layer
 */
export interface DOMAnalysis {
  element_count: number;
  div_ratio: number;
  semantic_score: number; // 0-7 (factual count of semantic elements)
  heading_hierarchy_valid: boolean;
  duplicate_ids?: number; // count of duplicate IDs (0 = good)
  inline_event_handlers?: number; // count of inline handlers (0 = good)
  avg_element_depth?: number; // average depth of elements

  // --- Expanded xray.py fields (all optional, additive) ---
  depth_max?: number;
  unique_tags?: number;
  deepest_path?: string[];
  element_map?: Record<string, number>;
  heading_counts?: { h1: number; h2: number; h3: number };
  empty_elements?: number;
  deprecated_tags?: string[];
  iframes?: { count: number; domains: string[] };
  head_signals?: {
    charset_present: boolean;
    viewport_present: boolean;
    favicon_present: boolean;
    preload_count: number;
  };
  content_ratios?: {
    html_size_kb: number;
    word_count_approx: number;
    html_text_ratio: number;
  };
  accessibility?: {
    images_missing_alt: number;
    inputs_without_label: number;
    buttons_links_no_text: number;
    html_missing_lang: boolean;
    aria_attribute_count: number;
  };
  links_summary?: {
    total: number;
    internal: number;
    external: number;
    target_blank_missing_rel: number;
  };
  images_summary?: {
    total: number;
    missing_alt: number;
    missing_dimensions: number;
    lazy_loaded: number;
  };
  forms_summary?: {
    form_count: number;
    input_count: number;
    button_count: number;
    inputs_without_labels: number;
    forms_missing_action: number;
  };
  scripts_summary?: {
    total: number;
    inline: number;
    external: number;
    defer_count: number;
    async_count: number;
    third_party_count: number;
    third_party_domains: string[];
    tag_manager_detected: boolean;
  };
  inline_styles_count?: number;

  // --- Phase 4 (webflow-tier) additions ---
  depth_avg?: number;
  tabindex_audit?: { positive_tabindex_count: number };
  largest_image_candidate?: { src: string; width: number; height: number } | null;
  text_density_by_region?: { main: number; aside: number; footer: number; header: number };
  duplicate_headings?: string[];
}

/**
 * Technical SEO analysis from Python layer
 */
export interface TechnicalSEO {
  title_present: boolean;
  description_present: boolean;
  canonical_present: boolean;
  schema_blocks: number;
  open_graph_present?: boolean;
  is_indexable?: boolean;
  twitter_card_present?: boolean;
  security_headers_count?: number;

  // --- Expanded technical_seo.py fields (all optional, additive) ---
  title?: { content?: string; length?: number; status?: string };
  description?: { content?: string; length?: number; status?: string };
  robots?: { index: boolean; follow: boolean; content?: string; status?: string };
  charset_present?: boolean;
  viewport_present?: boolean;
  canonical?: { href?: string | null; self_referencing?: boolean | null; status?: string };
  open_graph?: {
    title: boolean;
    description: boolean;
    image: boolean;
    url: boolean;
    published_time?: string | null;
    modified_time?: string | null;
    updated_time?: string | null;
  };
  indexability?: { blocked: boolean; signals: string[]; conflicts: string[] };
  links_summary?: {
    internal_total: number;
    internal_generic_anchor: number;
    external_total: number;
    external_broken: number;
  };
  security_headers?: {
    present: string[];
    missing: string[];
    count: number;
    grade?: string;
    details?: Record<string, unknown>;
  };
  hreflang?: {
    present: boolean;
    count: number;
    languages: string[];
    has_x_default: boolean;
    issues: string[];
  };
  pagination_amp?: {
    has_prev: boolean;
    has_next: boolean;
    prev_href?: string | null;
    next_href?: string | null;
    is_paginated: boolean;
    amp_link_present: boolean;
    amp_html?: string | null;
    is_amp: boolean;
  };
  caching?: {
    cache_control?: string | null;
    has_cache_control: boolean;
    has_etag: boolean;
    has_last_modified: boolean;
    max_age_seconds?: number | null;
    is_cacheable: boolean;
    issues: string[];
  };
  resource_hints?: {
    preload: string[];
    prefetch: string[];
    dns_prefetch: string[];
    preconnect: string[];
    counts: Record<string, number>;
  };
  url_structure?: {
    length: number;
    path: string;
    has_trailing_slash: boolean;
    has_uppercase: boolean;
    has_special_chars: boolean;
    has_double_slashes: boolean;
    keyword_segments: number;
    total_segments: number;
    issues: string[];
  };
}

/**
 * Content analysis from Python layer (Section 5)
 */
export interface ContentAnalysis {
  // Existing fields — MUST keep exact shape for backward compatibility.
  issues: string[];

  // Summary fields that merger.ts writes to report.content_analysis.
  // These live on the mapped object too so merger can read them directly.
  depth_label?: string;
  eeat_label?: string;
  freshness_status?: string;
  thin_content_risk?: string;
  anchor_quality_score?: string; // Python returns this as a label string ("excellent" | "good" | ...)
  snippet_eligible?: boolean;

  // --- New detail fields from content_analysis.py (all optional, additive) ---
  detected_language?: string;
  content_depth?: {
    word_count: number;
    paragraph_count: number;
    avg_paragraph_length: number;
    depth_label: string;
    reading_time_minutes?: number;
    lexical_diversity?: number;
    lexical_diversity_label?: string;
    paragraph_length_distribution?: { min: number; max: number; p50: number; p90: number };
    issues: string[];
  };
  content_relevance?: {
    title_in_h1: boolean;
    title_in_intro: boolean;
    title_in_intro_word_position?: number | null;
    heading_alignment_score: number;
    keyword_stuffing_detected: boolean;
  };
  eeat_signals?: {
    first_person_count: number;
    first_person_present: boolean;
    statistics_count: number;
    citation_patterns: number;
    author_mention_detected: boolean;
    eeat_label: string;
    eeat_signals_count: number;
    eeat_signals_present: Record<string, boolean>;
    dates_found: unknown[];
    most_recent_date?: string | number | null;
    time_sensitive_without_date: boolean;
  };
  content_freshness?: {
    years_mentioned: number[];
    most_recent_year?: number | null;
    current_year: number;
    freshness_status: string;
    time_sensitive_phrases_found: string[];
    time_sensitive_without_date: boolean;
  };
  featured_snippet?: {
    definition_paragraph_present: boolean;
    list_snippet_eligible: boolean;
    lists_under_headings: unknown[];
    qa_pairs_found: unknown[];
    qa_pattern_count: number;
    faq_schema_recommended: boolean;
    tables_with_headers: number;
    table_snippet_eligible: boolean;
    snippet_types_eligible: string[];
    snippet_eligible: boolean;
  };
  thin_content?: {
    boilerplate_detected: string[];
    boilerplate_present: boolean;
    duplicate_sentences_found: number;
    duplicate_paragraphs_found?: number;
    high_repetition: boolean;
    heading_count: number;
    heading_to_content_ratio: number;
    skeleton_page_detected: boolean;
    thin_content_signals: Record<string, boolean>;
    thin_content_risk: string;
  };
  anchor_text_quality?: {
    total_internal_links: number;
    descriptive_count: number;
    partial_count: number;
    generic_count: number;
    naked_url_count: number;
    empty_count: number;
    descriptive_ratio: number;
    anchor_quality_score: string;
  };
  readability?: {
    avg_words_per_sentence: number;
    long_sentences_count: number;
    short_sentences_count: number;
    flesch_reading_ease: number;
    gunning_fog_index: number;
    reading_level: string;
    sentence_count: number;
    sentence_length_distribution?: { min: number; max: number; p50: number; p90: number };
  };
  passive_voice?: {
    passive_voice_count: number;
    passive_voice_ratio: number;
    examples?: string[];
  };
  first_paragraph?: {
    word_count: number;
    contains_title_keyword: boolean;
    has_hook: boolean;
  };
  transition_words?: {
    transition_word_count: number;
    transition_word_ratio: number;
    transition_label: string;
  };
  meta_description_info?: {
    meta_description_length: number;
    meta_description_status: string;
  };
  link_density?: {
    links_per_1000_words: number;
    total_internal_links: number;
    issues: string[];
  };
  image_alt_text?: {
    images_total: number;
    images_missing_alt: number;
    alt_coverage_ratio: number;
    images_empty_alt?: number;
    images_decorative?: number;
    images_informative?: number;
  };
  heading_hierarchy?: {
    hierarchy_valid: boolean;
    violations: Array<Record<string, unknown>> | string[];
    h1_count?: number;
    skipped_levels?: number;
    orphan_headings?: number;
  };
  toc?: { toc_present: boolean; toc_entry_count: number; toc_recommended: boolean };
  cta?: { cta_present: boolean; cta_patterns_found: string[]; cta_count: number };
  author_bio?: { author_bio_present: boolean; detected_pattern?: string };
  top_keywords?: Array<{ word: string; count: number; percentage: number }>;
  top_phrases?: {
    bigrams: unknown[];
    trigrams: unknown[];
  };
}

/**
 * On-page SEO analysis from Python layer
 */
export interface OnPageSEO {
  h1_count: number;
  content_word_count: number;
  image_alt_missing: number;
  internal_links: number;
  heading_hierarchy_valid?: boolean;
  has_robots?: boolean;

  // --- Expanded onpage.py fields (all optional, additive) ---
  heading_counts?: { h1: number; h2: number; h3: number; h4: number; h5: number; h6: number };
  h1_content?: string[];
  empty_headings?: number;
  total_headings?: number;
  heading_violations?: Array<{
    from_level: number;
    to_level: number;
    heading?: string;
    issue_type: string;
  }>;
  heading_tree?: Array<{
    level: number;
    text: string;
    children: unknown[];
  }>;
  heading_issues?: string[];
  table_of_contents_detected?: boolean;
  content?: {
    word_count: number;
    paragraph_count: number;
    avg_paragraph_length: number;
  };
  links_detail?: {
    internal_total: number;
    internal_generic_anchor: number;
    external_total: number;
    external_broken: number;
  };
  images_detail?: {
    total: number;
    missing_alt: number;
    empty_alt_decorative: number;
    too_short: number;
    too_long: number;
    poor_quality_alt: number;
    lazy_loading: number;
    modern_format: number;
    explicit_dimensions: number;
    density_per_1000_words: number;
  };
}

/**
 * Sitemap analysis block (one entry per Sitemap: directive).
 */
export interface SitemapAnalysis {
  url?: string;
  url_count?: number;
  has_lastmod?: boolean;
  is_index?: boolean;
  error?: string | null;
  children_fetched?: number;
  total_urls_across_children?: number;
  discovered_via_fallback?: boolean;
}

/**
 * Multi-agent verdict: per-user-agent allowed/disallowed for the analyzed URL.
 */
export type AgentVerdict = 'allowed' | 'disallowed';

/**
 * AI-bot blocking summary — counts explicit blocks per AI crawler.
 */
export interface AIBotSummary {
  blocked_count: number;
  blocked_agents: string[];
  total_checked: number;
}

/**
 * Robots.txt analysis from robots_check.py
 */
export interface RobotsInfo {
  fetched?: boolean;
  status_code?: number;
  path_disallowed?: boolean;
  reason?: string;
  disallow_rules?: string[];
  allow_rules?: string[];
  crawl_delay?: number | null;
  sitemaps?: string[];
  has_wildcard_disallow?: boolean;
  conflict_with_meta?: boolean;
  issues?: string[];
  sitemap_analysis?: SitemapAnalysis;

  // --- Expanded robots_check.py fields (all optional, additive) ---
  final_url?: string | null;
  content_type?: string | null;
  content_length?: number | null;
  elapsed_ms?: number;
  redirect_chain?: string[];
  per_agent_rules?: Record<string, { disallow: string[]; allow: string[]; crawl_delay: number | null }>;
  per_agent_verdict?: Record<string, AgentVerdict>;
  ai_bot_summary?: AIBotSummary;
  sitemap_analyses?: SitemapAnalysis[];
  syntax_warnings?: string[];
  size_exceeds_google_limit?: boolean;
  content_type_is_text_plain?: boolean;
  cross_origin_redirect?: boolean;
}

/**
 * Schema.org validation block-level detail from schema_validator.py
 */
export interface SchemaValidationBlock {
  type?: string;
  raw_json?: unknown;
  score?: number;
  validation?: {
    required?: { fields: string[]; present: string[]; missing: string[] };
    recommended?: { fields: string[]; present: string[]; missing: string[] };
    format_errors?: Array<{
      field: string;
      value?: unknown;
      expected?: string;
      message: string;
    }>;
    warnings?: Array<{ field: string; message: string }>;
  };
  rich_results?: {
    eligible: boolean;
    types: string[];
    missing_for_eligibility: string[];
  };
}

/**
 * Schema validation summary + block detail from schema_validator.py
 */
export interface SchemaValidationInfo {
  blocks_found?: number;
  total_blocks?: number;
  valid_blocks?: number;
  types?: string[];
  rich_results_eligible?: string[];
  rich_results_ineligible?: string[];
  duplicate_types?: string[];
  overall_score?: number;
  recommendations?: Array<{ priority?: string; type?: string; message: string }>;
  blocks?: SchemaValidationBlock[];
}

/**
 * Aggregated analysis data from all phases
 */
export interface AnalysisData {
  fetch?: FetchResult | null;
  field_data?: FieldData | null;
  lab_data?: LabData | null;
  dom?: DOMAnalysis | null;
  technical_seo?: TechnicalSEO | null;
  onpage_seo?: OnPageSEO | null;
  content_analysis?: ContentAnalysis | null;
  error?: string;
}
